File: golang/20-legacy-monolith/legacy_system_test.go
package kata

import (
	"strings"
	"testing"
)

func TestProcessEverythingRejectsEmptyOrder(t *testing.T) {
	ls := NewLegacySystem()
	result := ls.process_everything(MonoOrder{ID: 1, Items: []MonoItem{}})
	if result.Error != "No items" {
		t.Errorf("expected error 'No items', got %q", result.Error)
	}
}

func TestProcessEverythingCalculatesBasicTotal(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 1, Items: []MonoItem{{Price: 10, Quantity: 2}}, Customer: MonoCustomer{Email: "a@b.com"}}
	result := ls.process_everything(order)
	if result.Total != 21.4 {
		t.Errorf("expected total 21.4, got %v", result.Total)
	}
	if result.Carrier != "USPS" {
		t.Errorf("expected carrier 'USPS', got %q", result.Carrier)
	}
}

func TestProcessEverythingAppliesMemberDiscount(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 2, Items: []MonoItem{{Price: 100, Quantity: 1}}, Customer: MonoCustomer{Type: "member", Email: "a@b.com"}}
	result := ls.process_everything(order)
	if result.Total != 96.3 {
		t.Errorf("expected total 96.3, got %v", result.Total)
	}
}

func TestProcessEverythingAppliesVipDiscount(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 3, Items: []MonoItem{{Price: 100, Quantity: 1}}, Customer: MonoCustomer{Type: "vip", Email: "a@b.com"}}
	result := ls.process_everything(order)
	if result.Total != 85.6 {
		t.Errorf("expected total 85.6, got %v", result.Total)
	}
}

func TestProcessEverythingAppliesBonusDiscountOver100(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 4, Items: []MonoItem{{Price: 200, Quantity: 1}}, Customer: MonoCustomer{Email: "a@b.com"}}
	result := ls.process_everything(order)
	if result.Total != 203.3 {
		t.Errorf("expected total 203.3, got %v", result.Total)
	}
}

func TestProcessEverythingUsesUPSForLargeTotal(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 5, Items: []MonoItem{{Price: 60, Quantity: 1}}, Customer: MonoCustomer{Email: "a@b.com"}}
	result := ls.process_everything(order)
	if result.Carrier != "UPS" {
		t.Errorf("expected carrier 'UPS', got %q", result.Carrier)
	}
}

func TestProcessEverythingFlagsHighTotalForReview(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 6, Items: []MonoItem{{Price: 5000, Quantity: 1}}, Customer: MonoCustomer{Email: "a@b.com"}}
	result := ls.process_everything(order)
	if result.PaymentStatus != "manual_review" {
		t.Errorf("expected payment status 'manual_review', got %q", result.PaymentStatus)
	}
}

func TestProcessEverythingIncludesEmailDetails(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 7, Items: []MonoItem{{Price: 10, Quantity: 1}}, Customer: MonoCustomer{Email: "user@test.com"}}
	result := ls.process_everything(order)
	if result.EmailTo != "user@test.com" {
		t.Errorf("expected email to 'user@test.com', got %q", result.EmailTo)
	}
}

func TestProcessEverythingIncludesLogEntry(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 8, Items: []MonoItem{{Price: 10, Quantity: 1}}, Customer: MonoCustomer{Email: "a@b.com"}}
	result := ls.process_everything(order)
	if !strings.Contains(result.Log, "Order processed") {
		t.Errorf("expected log to contain 'Order processed', got %q", result.Log)
	}
}

func TestProcessEverythingIncludesOrderId(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 99, Items: []MonoItem{{Price: 10, Quantity: 1}}, Customer: MonoCustomer{Email: "a@b.com"}}
	result := ls.process_everything(order)
	if result.OrderID != 99 {
		t.Errorf("expected order id 99, got %d", result.OrderID)
	}
}

func TestProcessEverythingIgnoresNonPositiveItemPricesButCountsPositiveQuantities(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 10, Items: []MonoItem{{Price: 20, Quantity: 2}, {Price: -100, Quantity: 1}}, Customer: MonoCustomer{Email: "a@b.com"}}
	result := ls.process_everything(order)
	if result.Total != 42.8 {
		t.Errorf("expected total 42.8, got %v", result.Total)
	}
	if result.ShippingWeight != 3 {
		t.Errorf("expected shipping weight 3, got %d", result.ShippingWeight)
	}
}

func TestProcessEverythingAppliesSave10Coupon(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 11, Items: []MonoItem{{Price: 100, Quantity: 1}}, Coupon: "SAVE10", Customer: MonoCustomer{Email: "a@b.com"}}
	result := ls.process_everything(order)
	if result.Total != 96.3 {
		t.Errorf("expected total 96.3, got %v", result.Total)
	}
}

func TestProcessEverythingTaxExemptCustomerPaysNoTax(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 12, Items: []MonoItem{{Price: 100, Quantity: 1}}, Customer: MonoCustomer{Email: "a@b.com", TaxExempt: true}}
	result := ls.process_everything(order)
	if result.Total != 100 {
		t.Errorf("expected total 100, got %v", result.Total)
	}
	if result.TaxRate != 0 {
		t.Errorf("expected tax rate 0, got %v", result.TaxRate)
	}
}

func TestProcessEverythingExpressShippingOverridesCarrier(t *testing.T) {
	ls := NewLegacySystem()
	order := MonoOrder{ID: 13, Items: []MonoItem{{Price: 10, Quantity: 1}}, Shipping: MonoShipping{Speed: "express"}, Customer: MonoCustomer{Email: "a@b.com"}}
	result := ls.process_everything(order)
	if result.Carrier != "FedEx" {
		t.Errorf("expected FedEx, got %q", result.Carrier)
	}
}

File: csharp/20-legacy-monolith/test/legacy_system_test.cs
using Xunit;

public class LegacySystemTest
{
    [Fact]
    public void RejectsEmptyOrder()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(1, new List<MonoItem>(), new MonoCustomer("", "")));
        Assert.Equal("No items", result.Error);
    }

    [Fact]
    public void CalculatesBasicTotal()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(1,
            new List<MonoItem> { new MonoItem(10, 2) },
            new MonoCustomer("", "a@b.com")));
        Assert.Equal(21.4m, result.Total);
        Assert.Equal("USPS", result.Carrier);
    }

    [Fact]
    public void AppliesMemberDiscount()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(2,
            new List<MonoItem> { new MonoItem(100, 1) },
            new MonoCustomer("member", "a@b.com")));
        Assert.Equal(96.3m, result.Total);
    }

    [Fact]
    public void AppliesVipDiscount()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(3,
            new List<MonoItem> { new MonoItem(100, 1) },
            new MonoCustomer("vip", "a@b.com")));
        Assert.Equal(85.6m, result.Total);
    }

    [Fact]
    public void AppliesBonusDiscountOver100()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(4,
            new List<MonoItem> { new MonoItem(200, 1) },
            new MonoCustomer("", "a@b.com")));
        Assert.Equal(203.3m, result.Total);
    }

    [Fact]
    public void UsesUpsForLargeTotal()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(5,
            new List<MonoItem> { new MonoItem(60, 1) },
            new MonoCustomer("", "a@b.com")));
        Assert.Equal("UPS", result.Carrier);
    }

    [Fact]
    public void FlagsHighTotalForReview()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(6,
            new List<MonoItem> { new MonoItem(5000, 1) },
            new MonoCustomer("", "a@b.com")));
        Assert.Equal("manual_review", result.PaymentStatus);
    }

    [Fact]
    public void IncludesEmailDetails()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(7,
            new List<MonoItem> { new MonoItem(10, 1) },
            new MonoCustomer("", "user@test.com")));
        Assert.Equal("user@test.com", result.Email.To);
        Assert.Contains("7", result.Email.Subject);
    }

    [Fact]
    public void IncludesLogEntry()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(8,
            new List<MonoItem> { new MonoItem(10, 1) },
            new MonoCustomer("", "a@b.com")));
        Assert.Contains("Order processed", result.Log);
    }

    [Fact]
    public void IncludesOrderId()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(99,
            new List<MonoItem> { new MonoItem(10, 1) },
            new MonoCustomer("", "a@b.com")));
        Assert.Equal(99, result.OrderId);
    }

    [Fact]
    public void IgnoresNonPositiveItemPricesButCountsPositiveQuantities()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(10,
            new List<MonoItem> { new MonoItem(20, 2), new MonoItem(-100, 1) },
            new MonoCustomer("", "a@b.com")));
        Assert.Equal(42.8m, result.Total);
        Assert.Equal(3, result.ShippingWeight);
    }

    [Fact]
    public void AppliesSave10Coupon()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(11,
            new List<MonoItem> { new MonoItem(100, 1) },
            new MonoCustomer("", "a@b.com"),
            "SAVE10"));
        Assert.Equal(96.3m, result.Total);
    }

    [Fact]
    public void TaxExemptCustomerPaysNoTax()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(12,
            new List<MonoItem> { new MonoItem(100, 1) },
            new MonoCustomer("", "a@b.com", TaxExempt: true)));
        Assert.Equal(100m, result.Total);
        Assert.Equal(0m, result.TaxRate);
    }

    [Fact]
    public void ExpressShippingOverridesCarrier()
    {
        var system = new LegacySystem();
        var result = system.process_everything(new MonoOrder(13,
            new List<MonoItem> { new MonoItem(10, 1) },
            new MonoCustomer("", "a@b.com"),
            Shipping: new MonoShipping("express")));
        Assert.Equal("FedEx", result.Carrier);
    }
}
